synse server error when no reading returned from sdk

this is for sdk 1.0

```
synse-server    | [2018-06-19 18:10:41 +0000] - (synse)[WARNING] read:84: Read for ipmi/ipmisim/cf5ca636df49458db117b2dd2e9fb1ec returned gRPC "no readings found". Will apply None as reading value. Note that this response might indicate plugin error/misconfiguration.
ipmi-plugin     | time="2018-06-19T18:10:41Z" level=debug msg="[grpc] read rpc request" request="rack:\"ipmi\" board:\"ipmisim\" device:\"cf5ca636df49458db117b2dd2e9fb1ec\" "
ipmi-plugin     | time="2018-06-19T18:10:41Z" level=error msg="[data manager] no readings found" id=ipmi-ipmisim-cf5ca636df49458db117b2dd2e9fb1ec
synse-server    | [2018-06-19 18:10:42 +0000] - (root)[ERROR] handlers:105: Traceback (most recent call last):
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/commands/read.py", line 41, in read
synse-server    |     read_data = _plugin.client.read(rack, board, device)
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/proto/client.py", line 188, in read
synse-server    |     resp = [r for r in self.grpc.Read(req, timeout=timeout)]
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/proto/client.py", line 188, in <listcomp>
synse-server    |     resp = [r for r in self.grpc.Read(req, timeout=timeout)]
synse-server    |   File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 347, in __next__
synse-server    |     return self._next()
synse-server    |   File "/usr/local/lib/python3.6/site-packages/grpc/_channel.py", line 341, in _next
synse-server    |     raise self
synse-server    | grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.NOT_FOUND, no readings found for device: ipmi-ipmisim-cf5ca636df49458db117b2dd2e9fb1ec)>
synse-server    | 
synse-server    | During handling of the above exception, another exception occurred:
synse-server    | 
synse-server    | Traceback (most recent call last):
synse-server    |   File "/usr/local/lib/python3.6/site-packages/google/protobuf/internal/python_message.py", line 545, in _GetFieldByName
synse-server    |     return message_descriptor.fields_by_name[field_name]
synse-server    | KeyError: 'value'
synse-server    | 
synse-server    | During handling of the above exception, another exception occurred:
synse-server    | 
synse-server    | Traceback (most recent call last):
synse-server    |   File "/usr/local/lib/python3.6/site-packages/sanic/app.py", line 556, in handle_request
synse-server    |     response = await response
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/validate.py", line 85, in inner
synse-server    |     return await f(request, *args, **kwargs)
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/routes/core.py", line 72, in read_route
synse-server    |     response = await commands.read(rack, board, device)
synse-server    |   File "/usr/local/lib/python3.6/site-packages/synse-2.1.0-py3.6.egg/synse/commands/read.py", line 92, in read
synse-server    |     value='',
synse-server    |   File "/usr/local/lib/python3.6/site-packages/google/protobuf/internal/python_message.py", line 484, in init
synse-server    |     field = _GetFieldByName(message_descriptor, field_name)
synse-server    |   File "/usr/local/lib/python3.6/site-packages/google/protobuf/internal/python_message.py", line 548, in _GetFieldByName
synse-server    |     (message_descriptor.name, field_name))
synse-server    | ValueError: Protocol message Reading has no "value" field.
synse-server    | 
```

--- a/synse/commands/read.py
+++ b/synse/commands/read.py
@@ -35,7 +35,6 @@ async def read(rack, board, device):
             _('Unable to find plugin named "{}" to read').format(plugin_name)
         )
 
-    read_data = []
     try:
         # Perform a gRPC read on the device's managing plugin
         read_data = _plugin.client.read(rack, board, device)
@@ -47,6 +46,11 @@ async def read(rack, board, device):
         # is a better way to check this other than comparing strings..
         if hasattr(ex, 'code') and hasattr(ex, 'details'):
             if grpc.StatusCode.NOT_FOUND == ex.code() and 'no readings found' in ex.details().lower():
+
+                # FIXME (etd) - with SDK v1.0, is the below correct? We should not longer
+                # have to pass the "null" string. I think an empty string should also not
+                # indicate no readings.. it should be the NOT_FOUND error (or, at least
+                # some kind of error).
 
                 # Currently, in the SDK, there are three different behaviors for
                 # devices that do not have readings. Either (a). "null" is returned,
@@ -89,9 +93,10 @@ async def read(rack, board, device):
                         api.Reading(
                             timestamp=utils.rfc3339now(),
                             type=output.type,
-                            value='',
                         )
                     )
+            else:
+                raise errors.FailedReadCommandError(str(ex)) from ex
         else:
             raise errors.FailedReadCommandError(str(ex)) from ex
 

--- a/tests/unit/commands/test_read_command.py
+++ b/tests/unit/commands/test_read_command.py
@@ -1,5 +1,5 @@
 """Test the 'synse.commands.read' Synse Server module."""
-# pylint: disable=redefined-outer-name,unused-argument
+# pylint: disable=redefined-outer-name,unused-argument,line-too-long
 
 import os
 import shutil
@@ -10,7 +10,7 @@ import pytest
 from synse_grpc import api
 
 import synse.cache
-from synse import errors, plugin
+from synse import errors, plugin, utils
 from synse.commands.read import read
 from synse.proto.client import PluginClient, PluginUnixClient
 from synse.scheme.read import ReadResponse
@@ -78,6 +78,15 @@ def mockreadfail(self, rack, board, device):
     raise grpc.RpcError()
 
 
+def patch_client_read_error(self, rack, board, device):
+    """Patch the grpc client's read method to raise a gRPC error indicative of
+    the "no readings found" case."""
+    e = grpc.RpcError()
+    e.details = lambda: 'no readings found'
+    e.code = lambda: grpc.StatusCode.NOT_FOUND
+    raise e
+
+
 @pytest.fixture()
 def mock_get_device_info(monkeypatch):
     """Fixture to monkeypatch the cache device meta lookup."""
@@ -98,6 +107,22 @@ def mock_client_read_fail(monkeypatch):
     """Fixture to monkeypatch the grpc client's read method to fail."""
     monkeypatch.setattr(PluginClient, 'read', mockreadfail)
     return mock_client_read_fail
+
+
+@pytest.fixture()
+def mock_client_read_error(monkeypatch):
+    """Fixture to monkeypatch the client's read method to error out."""
+    monkeypatch.setattr(PluginClient, 'read', patch_client_read_error)
+    return mock_client_read_error
+
+
+@pytest.fixture()
+def mock_rfc_time(monkeypatch):
+    """Fixture to monkeypatch the util for generating a timestamp so we can
+    reliably test the timestamp field.
+    """
+    monkeypatch.setattr(utils, 'rfc3339now', lambda: 'sometime')
+    return mock_rfc_time
 
 
 @pytest.fixture()
@@ -157,6 +182,30 @@ async def test_read_command_grpc_err(mock_get_device_info, mock_client_read_fail
         await read('rack-1', 'vec', '12345')
     except errors.SynseError as e:
         assert e.error_id == errors.FAILED_READ_COMMAND
+
+
+@pytest.mark.asyncio
+async def test_read_command_grpc_err_no_reading(mock_get_device_info, mock_client_read_error, make_plugin, mock_rfc_time):
+    """Get a ReadResponse when the plugin returns an error indicating no reading."""
+
+    resp = await read('rack-1', 'vec', '12345')
+
+    assert isinstance(resp, ReadResponse)
+    assert resp.data == {
+        'kind': 'thermistor',
+        'data': {
+            'temperature': {
+                'info': '',
+                'type': 'temperature',
+                'value': None,
+                'timestamp': 'sometime',
+                'unit': {
+                    'name': 'celsius',
+                    'symbol': 'C'
+                }
+            }
+        }
+    }
 
 
 @pytest.mark.asyncio

